### PR TITLE
timeout version of esdm_get_random_bytes_full

### PR DESCRIPTION
**Pull request by smuellerDD**
As the esdm_get_random_bytes_full may sleep indefinitely if the ESDM is not fully seeded at the time the entropy sources cannot provide entropy, allow the caller to specify a timeout with the maximum waiting time. Once the timeout elapsed, the esdm_get_random_bytes_full_timeout returns with an error -ETIMEDOUT.

Also provide the RPC wrappers around this call.

### CODE REVIEW

**Comment by thillux** [comment]
Hi Stephan,

I did a quick check of this branch and never managed to get random bytes from esdm_rpcc_get_random_bytes_full_timeout. Even when esdm-server is fully seeded, I got -22 (-EINVAL) from this function.

My test code looks like:

```c++
#include <ctime>
#include <esdm/esdm_rpc_client.h>
#include <time.h>
#include <cassert>
#include <iostream>

int main(void) {
    int ret = esdm_rpcc_init_unpriv_service(nullptr);
    assert(ret == 0);

    const size_t BUF_SIZE = 32;
    uint8_t buf[BUF_SIZE];

    struct timespec ts;
    ret = clock_gettime(CLOCK_MONOTONIC, &ts);
    ts.tv_sec += 10;
    assert(ret == 0);

    ret = esdm_rpcc_get_random_bytes_full_timeout(buf, BUF_SIZE, &ts);
    std::cout << "return value: " << ret << std::endl;

    esdm_rpcc_fini_unpriv_service();
}
```

Your code currently uses a CLOCK_REALTIME based clock, I'd like to change this to CLOCK_MONOTONIC (also tested with CLOCK_REALTIME, but got -22 there too). I'll open a PR to change the clock behavior later this day.

BR
Markus

File: service-rpc/client/esdm_rpc_get_random_bytes_full_timeout.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+out:
+	esdm_rpcc_put_unpriv_service(rpc_conn);
+	return (ret < 0) ? ret : (ssize_t)orig_buflen;
+}
+
+DSO_PUBLIC
+ssize_t esdm_rpcc_get_random_bytes_full_timeout(uint8_t *buf, size_t buflen,
+						struct timespec *ts)
+{
+	return esdm_rpcc_get_random_bytes_full_timeout_int(buf, buflen, NULL,
```
Comment (myrkr) [inline service-rpc/client/esdm_rpc_get_random_bytes_full_timeout.c]: The parameters for timespec and int_data are in the wrong order, causing -EINVAL on every call through esdm_rpcc_get_random_bytes_full_timeout

Comment (myrkr) [reply service-rpc/client/esdm_rpc_get_random_bytes_full_timeout.c]: After fixing this issues first tests looked good, the timeout handling was working correctly and I also did some tests with concurrent calls to that functions.

**Comment by smuellerDD** [comment]
Am Montag, 12. August 2024, 09:23:40 MESZ schrieb Torsten Hilbrich:

Hi Torsten,

> After fixing this issues first tests looked good, the timeout handling was
> working correctly and I also did some tests with concurrent calls to that
> functions.

Sorry for the error - the compiler did not complain due to using a void 
pointer.

I pushed the update to the branch. Once your testing is complete, please let 
me know if I can merge the patch.

Ciao
Stephan

**Comment by smuellerDD** [comment]
Am Montag, 12. August 2024, 08:17:46 MESZ schrieb Markus Theil:

Hi Markus,

> Your code currently uses a CLOCK_REALTIME based clock, I'd like to change
> this to CLOCK_MONOTONIC (also tested with CLOCK_REALTIME, but got -22 there
> too). I'll open a PR to change the clock behavior later this day.

I think you refer to thread_timedwait_no_event?

Just open an issue. Since this is separate to the timeout API, let us handle 
it separately.

Thanks

Ciao
Stephan

**Comment by myrkr** [comment]
> Am Montag, 12. August 2024, 09:23:40 MESZ schrieb Torsten Hilbrich: Hi Torsten,
> After fixing this issues first tests looked good, the timeout handling was working correctly and I also did some tests with concurrent calls to that functions.
> Sorry for the error - the compiler did not complain due to using a void pointer. I pushed the update to the branch. Once your testing is complete, please let me know if I can merge the patch. Ciao Stephan

IMO the code can now be merged. I did more testing with the fixed code and found no problem so far.

**Comment by thillux** [comment]
> Am Montag, 12. August 2024, 08:17:46 MESZ schrieb Markus Theil: Hi Markus,
> Your code currently uses a CLOCK_REALTIME based clock, I'd like to change this to CLOCK_MONOTONIC (also tested with CLOCK_REALTIME, but got -22 there too). I'll open a PR to change the clock behavior later this day.
> I think you refer to thread_timedwait_no_event? Just open an issue. Since this is separate to the timeout API, let us handle it separately. Thanks Ciao Stephan

Hi Stephan,

sounds good. :+1: 

BR
Markus

**Pull request closed without merging**